RHDEVDOCS-4305: fix priorityClassName issue with code in modules/op-s…

ekristova opened the pull request:
• **Aligned team**: Dev Tools
• **OCP version for cherry-picking**: enterprise-4.11 and later
• **JIRA issue**: [RHDEVDOCS-4305](https://issues.redhat.com/browse/RHDEVDOCS-4305)
• **Preview page**: [Specifying pipelines resource quota using priority class](https://57466--docspreview.netlify.app/openshift-enterprise/latest/cicd/pipelines/setting-compute-resource-quota-for-openshift-pipelines.html#specifying-pipelines-resource-quota-using-priority-class_setting-compute-resource-quota-for-openshift-pipelines)
• **SME Review**: Completed by @vdemeester @concaf 
• **QE review**: Completed by @ppitonak 
• **Peer-review**: Completed by @mramendi

ekristova commented:
Hello @ppitonak @vdemeester . This is a PR raised to address the bug in https://github.com/openshift/openshift-docs/pull/48576#discussion_r936317926 . I have created this PR based on the previous writer's work. Please let me know if there is something missing. Thank you!

ppitonak commented on modules/op-specifying-pipelines-resource-quota-using-priority-class.adoc:
```diff
@@ -167,13 +166,16 @@ Ensure that all tasks in the pipeline belongs to the same priority class.
 kind: PipelineRun
 metadata:
   generateName: petclinic-run-
 spec:
```
This a doc about how to use priority class but its' not defined anywhere. I tried to specify it in pipeline run
```
spec:
  podTemplate:
    priorityClassName: pipeline1-pc
```

but the run failed with following error. 

```
failed to create task run pod "petclinic-run-r7m29-build": pods "petclinic-run-r7m29-build-pod" is forbidden: failed quota: pipeline1-rq: must specify cpu for: prepare,step-git-source-source-dqghc,step-mvn,working-dir-initializer; memory for: prepare,step-git-source-source-dqghc,step-mvn,working-dir-initializer. Maybe missing or invalid Task asdf/mvn
```

Any idea how to fix it @vdemeester?

vdemeester replied:
@ppitonak the error you get is about a quota error though 🤔

vdemeester replied:
https://tekton.dev/docs/pipelines/podtemplates/ `priorityClass` is defined there though 🤔

ekristova replied:
@vdemeester The note with an error @ppitonak mentioned is right before the chapter we are working with: ([Alternative approaches for limiting compute resource consumption in OpenShift Pipelines](https://57466--docspreview.netlify.app/openshift-enterprise/latest/cicd/pipelines/setting-compute-resource-quota-for-openshift-pipelines.html#alternative-approaches-compute-resource-quota-pipelines_setting-compute-resource-quota-for-openshift-pipelines)). I wonder what would be the best approach here. Maybe to move the note to the procedure so that it is more visible to the user?  Or should we change the example ResourceQuota to include a limit range (if I understand correctly that the limit range is a solution here)?  WDYT?

vdemeester replied:
I am not entirely sure I follow 🤔 😅 
cc @concaf

ppitonak replied:
This document is talking about how to limit resource quota used by pipeline run.

1. the current version *does not do anything*
2. the document says that we would achieve the goal using priority class but priority class is not defined anywhere
3. when I define priority class, it fails - IMHO caused by the fact that we are missing range limit
4. I think that priority class is not the correct feature to achieve this goal

concaf replied:
@ppitonak i dug up my proposal doc which i wrote while adding these docs in downstream; where you ask that why is LimitRange not enough i found this:

> Why can we not use LimitRange?
A LimitRange can put constraints on pods or individual containers in a given namespace - but the scope is the entire namespace; there is no way to configure resources for a subset of pods in a namespace. So, it cannot distinguish pods running for a Pipeline from the rest of the pods.
For example, a Maven build task in a Pipeline requires much more resources than the pod for the resulting application. App operators need a way to be able to give enough quota for the Maven build for that application, without giving too much quota to all pods in that namespace.

> LimitRange applies to all pods and does not distinguish between Tekton and other pods. If a Pipeline runs resource-intensive builds, the customer has to increase the limit for all pods via the LimitRange in that namespace which is not acceptable for the customer.

concaf replied:
here is the doc btw - https://docs.google.com/document/d/1WTnkhI8wkn2rY7q_5q8BCGG9JD1_qPLiL2Y2ryOmHc4/edit

concaf commented on modules/op-specifying-pipelines-resource-quota-using-priority-class.adoc:
```diff
@@ -130,31 +130,30 @@ spec:
       args:
         - -Dmaven.repo.local=$(workspaces.maven-repo.path)
-        - "$(inputs.params.GOALS)"
-      priorityClassName: pipeline1-pc
```
why was this line removed? this causes the priority class to _not_ get applied to this task

concaf replied:
ah, i understand 😞
we need to set it in the pipelinerun then, and also set the limitrange 🤔
@vdemeester @ppitonak let's add it to the pipelinerun and also mention that users need to either set limitrange or specify resources explicitly, wdyt?
users need to play with both, limitrange and priority class to make this work.

vdemeester replied:
yes.

ekristova commented on modules/op-specifying-pipelines-resource-quota-using-priority-class.adoc:
```diff
@@ -130,50 +130,49 @@ spec:
+        - "$(params.GOALS)"
 ----
-+
-[NOTE]
-====
-Ensure that all tasks in the pipeline belongs to the same priority class.
-====
```
@concaf @vdemeester @ppitonak I removed this note since we do not specify the `priorityClassName` field in the tasks, but in the pipeline run. So, this note no longer made sense to me.

ekristova commented:
@concaf @vdemeester @ppitonak I added the required `priorityClassName` field in the pipeline run and also added the note. PTAL and let me know, if the procedure seems okay now. Thank you!

vdemeester reviewed (approved):
LGTM

ppitonak commented:
@vdemeester @concaf have you actually ran the example from this doc before approving it? It still fails on my cluster (OpenShift 4.12.12, Pipelines 1.10.0) with the following message
```
failed to create task run pod "petclinic-run-vssjb-build": pods "petclinic-run-vssjb-build-pod" is forbidden: failed quota: pipeline1-rq: must specify cpu for: prepare,step-git-source-source-hlb5f,step-mvn,working-dir-initializer; memory for: prepare,step-git-source-source-hlb5f,step-mvn,working-dir-initializer. Maybe missing or invalid Task testpc/mvn
```

ekristova commented:
> @vdemeester @concaf have you actually ran the example from this doc before approving it? It still fails on my cluster (OpenShift 4.12.12, Pipelines 1.10.0) with the following message
> 
> ```
> failed to create task run pod "petclinic-run-vssjb-build": pods "petclinic-run-vssjb-build-pod" is forbidden: failed quota: pipeline1-rq: must specify cpu for: prepare,step-git-source-source-hlb5f,step-mvn,working-dir-initializer; memory for: prepare,step-git-source-source-hlb5f,step-mvn,working-dir-initializer. Maybe missing or invalid Task testpc/mvn
> ```
Hello @ppitonak please correct me if I am wrong, but I thought that is why I was putting the note in the step to let users know about this issue and to guide them to steps to avoid that error. With that we can avoid putting that directly in the example, since the problem is not directly connected to this procedure and has 2 different solutions.

It is discussed in this thread: https://github.com/openshift/openshift-docs/pull/57466#discussion_r1164260646
Let me know, what you think and if you think different solution would be better. Thank you!

ppitonak reviewed (approved):
@ekristova Oh, my bad...I didn't notice this note. I'm OK having it this way.

ekristova commented:
@mramendi I incorporated your suggestions. Please let me know if everything looks good now. Thank you!

mramendi commented:
/lgtm

rolfedh commented:
/cherrypick enterprise-4.11

rolfedh commented:
/cherrypick enterprise-4.12

rolfedh commented:
/cherrypick enterprise-4.13
